File: uni_modules/uts-any-rtc/utssdk/app-android/RtcEventProxy.kt
package uts.sdk.modules.utsAnyRtc

import com.ss.bytertc.engine.handler.IRTCEngineEventHandler
import com.ss.bytertc.engine.handler.IRTCRoomEventHandler
import com.ss.bytertc.engine.data.RemoteStreamKey
import com.ss.bytertc.engine.data.StreamIndex
import com.ss.bytertc.engine.data.VideoFrameInfo
import com.ss.bytertc.engine.type.MediaStreamType
import com.ss.bytertc.engine.type.RTCRoomStats
import com.ss.bytertc.engine.type.StreamRemoveReason

class RtcEventProxy : IRTCEngineEventHandler() {

    // Engine event callbacks
    var onWarningCallback: ((code: Int, message: String) -> Unit)? = null
    var onErrorCallback: ((code: Int, message: String) -> Unit)? = null
    var onFirstRemoteVideoFrameDecodedCallback: ((uid: String, streamIndex: String, width: Int, height: Int) -> Unit)? = null

    override fun onWarning(warn: Int) {
        onWarningCallback?.invoke(warn, "Warning code: $warn")
    }

    override fun onError(err: Int) {
        onErrorCallback?.invoke(err, "Error code: $err")
    }

    override fun onFirstRemoteVideoFrameDecoded(remoteStreamKey: RemoteStreamKey, frameInfo: VideoFrameInfo) {
        val uid = remoteStreamKey.userId ?: ""
        val streamIndex = when (remoteStreamKey.streamIndex) {
            StreamIndex.STREAM_INDEX_SCREEN -> "screen"
            else -> "main"
        }
        onFirstRemoteVideoFrameDecodedCallback?.invoke(uid, streamIndex, frameInfo.width, frameInfo.height)
    }

    fun clearEngineCallbacks() {
        onWarningCallback = null
        onErrorCallback = null
        onFirstRemoteVideoFrameDecodedCallback = null
    }

    // Room event handler — one per room
    class RoomEventHandler(private val roomId: String) : IRTCRoomEventHandler() {

        var onRoomStateChangedCallback: ((roomId: String, uid: String, state: Int, extraInfo: String) -> Unit)? = null
        var onUserJoinedCallback: ((roomId: String, uid: String) -> Unit)? = null
        var onUserLeaveCallback: ((roomId: String, uid: String, reason: Int) -> Unit)? = null
        var onUserPublishStreamCallback: ((roomId: String, uid: String, mediaType: String) -> Unit)? = null
        var onUserUnpublishStreamCallback: ((roomId: String, uid: String, mediaType: String) -> Unit)? = null
        var onRoomStatsCallback: ((roomId: String, userCount: Int, duration: Int) -> Unit)? = null

        override fun onRoomStateChanged(roomId: String, uid: String, state: Int, extraInfo: String) {
            onRoomStateChangedCallback?.invoke(roomId, uid, state, extraInfo)
        }

        override fun onUserJoined(userInfo: com.ss.bytertc.engine.UserInfo) {
            onUserJoinedCallback?.invoke(roomId, userInfo.uid)
        }

        override fun onUserLeave(uid: String, reason: Int) {
            onUserLeaveCallback?.invoke(roomId, uid, reason)
        }

        override fun onUserPublishStream(uid: String, type: MediaStreamType) {
            onUserPublishStreamCallback?.invoke(roomId, uid, "main")
        }

        override fun onUserUnpublishStream(uid: String, type: MediaStreamType, reason: StreamRemoveReason) {
            onUserUnpublishStreamCallback?.invoke(roomId, uid, "main")
        }

        override fun onRoomStats(stats: RTCRoomStats?) {
            stats?.let {
                onRoomStatsCallback?.invoke(roomId, it.users, it.totalDuration)
            }
        }

        fun clearCallbacks() {
            onRoomStateChangedCallback = null
            onUserJoinedCallback = null
            onUserLeaveCallback = null
            onUserPublishStreamCallback = null
            onUserUnpublishStreamCallback = null
            onRoomStatsCallback = null
        }
    }

    companion object {
        private val roomHandlers: MutableMap<String, RoomEventHandler> = mutableMapOf()

        fun getOrCreateRoomHandler(roomId: String): RoomEventHandler {
            return roomHandlers.getOrPut(roomId) { RoomEventHandler(roomId) }
        }

        fun removeRoomHandler(roomId: String) {
            roomHandlers[roomId]?.clearCallbacks()
            roomHandlers.remove(roomId)
        }

        fun clearAllRoomHandlers() {
            roomHandlers.values.forEach { it.clearCallbacks() }
            roomHandlers.clear()
        }
    }
}

File: uni_modules/uts-any-rtc/utssdk/app-ios/RtcEventProxy.swift
import Foundation
import BytePlusRTC

// MARK: - Engine Event Proxy

class RtcEventProxy: NSObject, ByteRTCEngineDelegate {

    var onWarningCallback: ((_ code: NSNumber, _ message: String) -> Void)?
    var onErrorCallback: ((_ code: NSNumber, _ message: String) -> Void)?
    var onLocalVideoStatsCallback: ((_ width: Int, _ height: Int, _ sentKBitrate: Int, _ frameRate: Int) -> Void)?
    var onLocalAudioStatsCallback: ((_ sentKBitrate: Int, _ audioLevel: Int) -> Void)?

    func rtcEngine(_ engine: ByteRTCEngine, onWarning code: ByteRTCWarningCode) {
        onWarningCallback?(NSNumber(value: code.rawValue), "Warning code: \(code.rawValue)")
    }

    func rtcEngine(_ engine: ByteRTCEngine, onError code: ByteRTCErrorCode) {
        onErrorCallback?(NSNumber(value: code.rawValue), "Error code: \(code.rawValue)")
    }

    func rtcEngine(_ engine: ByteRTCEngine, onLocalVideoStats stats: ByteRTCLocalVideoStats) {
        onLocalVideoStatsCallback?(
            Int(stats.encodedFrameWidth),
            Int(stats.encodedFrameHeight),
            Int(stats.sentKBitrate),
            Int(stats.encoderOutputFrameRate)
        )
    }

    func rtcEngine(_ engine: ByteRTCEngine, onLocalAudioStats stats: ByteRTCLocalAudioStats) {
        onLocalAudioStatsCallback?(
            Int(stats.sentKBitrate),
            0
        )
    }

    func clearEngineCallbacks() {
        onWarningCallback = nil
        onErrorCallback = nil
        onLocalVideoStatsCallback = nil
        onLocalAudioStatsCallback = nil
    }
}

// MARK: - Room Event Proxy

class RtcRoomEventProxy: NSObject, ByteRTCRoomDelegate {

    private let roomId: String

    var onRoomStateChangedCallback: ((_ roomId: String, _ uid: String, _ state: NSNumber, _ extraInfo: String) -> Void)?
    var onUserJoinedCallback: ((_ roomId: String, _ uid: String, _ elapsed: NSNumber) -> Void)?
    var onUserLeaveCallback: ((_ roomId: String, _ uid: String, _ reason: NSNumber) -> Void)?
    var onUserPublishStreamCallback: ((_ roomId: String, _ uid: String, _ streamIndex: String) -> Void)?
    var onUserUnpublishStreamCallback: ((_ roomId: String, _ uid: String, _ streamIndex: String) -> Void)?
    var onFirstRemoteVideoFrameDecodedCallback: ((_ roomId: String, _ uid: String, _ streamIndex: String, _ width: NSNumber, _ height: NSNumber, _ elapsed: NSNumber) -> Void)?
    var onRoomStatsCallback: ((_ roomId: String, _ userCount: NSNumber, _ duration: NSNumber) -> Void)?

    init(roomId: String) {
        self.roomId = roomId
        super.init()
    }

    func rtcRoom(_ rtcRoom: ByteRTCRoom, onRoomStateChanged roomId: String, withUid uid: String, state: Int, extraInfo: String) {
        onRoomStateChangedCallback?(roomId, uid, NSNumber(value: state), extraInfo)
    }

    func rtcRoom(_ rtcRoom: ByteRTCRoom, onUserJoined userInfo: ByteRTCUserInfo) {
        onUserJoinedCallback?(roomId, userInfo.userId ?? "", NSNumber(value: 0))
    }

    func rtcRoom(_ rtcRoom: ByteRTCRoom, onUserLeave uid: String, reason: ByteRTCUserOfflineReason) {
        onUserLeaveCallback?(roomId, uid, NSNumber(value: reason.rawValue))
    }

    func rtcRoom(_ rtcRoom: ByteRTCRoom, onUserPublishStream streamKey: ByteRTCRemoteStreamKey, isScreen: Bool, mediaStreamType type: ByteRTCMediaStreamType) {
        let uid = streamKey.userId ?? ""
        onUserPublishStreamCallback?(roomId, uid, isScreen ? "screen" : "main")
    }

    func rtcRoom(_ rtcRoom: ByteRTCRoom, onUserUnpublishStream streamKey: ByteRTCRemoteStreamKey, mediaStreamType type: ByteRTCMediaStreamType, reason: ByteRTCStreamRemoveReason) {
        let uid = streamKey.userId ?? ""
        onUserUnpublishStreamCallback?(roomId, uid, isScreen(streamKey) ? "screen" : "main")
    }

    private func isScreen(_ key: ByteRTCRemoteStreamKey) -> Bool {
        return key.streamIndex == .indexScreen
    }

    func rtcRoom(_ rtcRoom: ByteRTCRoom, onRoomStats stats: ByteRTCRoomStats) {
        onRoomStatsCallback?(roomId, NSNumber(value: stats.userCount), NSNumber(value: stats.duration))
    }

    func rtcRoom(_ rtcRoom: ByteRTCRoom, onFirstRemoteVideoFrameDecoded remoteStreamKey: ByteRTCRemoteStreamKey, withFrameInfo frameInfo: ByteRTCVideoFrameInfo) {
        let uid = remoteStreamKey.userId ?? ""
        let streamIndex: String
        switch remoteStreamKey.streamIndex {
        case .indexScreen:
            streamIndex = "screen"
        default:
            streamIndex = "main"
        }
        onFirstRemoteVideoFrameDecodedCallback?(roomId, uid, streamIndex, NSNumber(value: frameInfo.width), NSNumber(value: frameInfo.height), NSNumber(value: 0))
    }

    func clearCallbacks() {
        onRoomStateChangedCallback = nil
        onUserJoinedCallback = nil
        onUserLeaveCallback = nil
        onUserPublishStreamCallback = nil
        onUserUnpublishStreamCallback = nil
        onFirstRemoteVideoFrameDecodedCallback = nil
        onRoomStatsCallback = nil
    }

    // MARK: - Static management

    private static var roomProxies: [String: RtcRoomEventProxy] = [:]

    static func getOrCreate(roomId: String) -> RtcRoomEventProxy {
        if let existing = roomProxies[roomId] {
            return existing
        }
        let proxy = RtcRoomEventProxy(roomId: roomId)
        roomProxies[roomId] = proxy
        return proxy
    }

    static func remove(roomId: String) {
        roomProxies[roomId]?.clearCallbacks()
        roomProxies.removeValue(forKey: roomId)
    }

    static func clearAll() {
        for (_, proxy) in roomProxies {
            proxy.clearCallbacks()
        }
        roomProxies.removeAll()
    }
}
